### ISSUE TITLE
number of classes does not match with CNN model 

### ISSUE THREAD

**Issue by asmaharry** [deepmedic/deepmedic#192]
hi,
In Brats 2018 dataset labels are 1,2 and 4 for three different tumors. so total number of classes including the background are 4. if i set the classification classes to 4 it gives me an error like number of classes does not match with CNN model.
then i checked it with 5 classes now it is running but for class 3 gives me warning at the time of sampling
"WARN: Invalid sampling category! Sampling map just zeros! No [Class-3] samples from this subject! "
and
validation for class 3 shows
VALIDATION: Epoch #0, Subepoch #3, Class-3: mean accuracy:   1.0000 => (TruePos+TrueNeg)/All-Predicted-Voxels = 5000/5000
VALIDATION: Epoch #0, Subepoch #3, Class-3: mean sensitivity: N/A => TruePos/RealPos = 0/0
VALIDATION: Epoch #0, Subepoch #3, Class-3: mean precision: N/A => TruePos/(TruePos+FalsePos) = 0/0
VALIDATION: Epoch #0, Subepoch #3, Class-3: mean specificity: 1.0000 => TrueNeg/RealNeg = 5000/5000
VALIDATION: Epoch #0, Subepoch #3, Class-3: mean Dice:       N/A

how i can solve this issue.

**Comment by Kamnitsask** [comment]
This kinda "works as intended". Let me explain:

The code sees you have a "4" in your ground truth labels.
So, it expects "at least" the classes 0,1,2,3,4. This makes it require number_of_classes=5 in the modelConfig.cfg.

The code creates an output layer with 5 output feature maps.

The class "3" in this case does not exist anywhere in the ground truth labels, hence you get a warning from the sampler.

But things will continue like normal.

The final "result" will be that the 4th feature map (the one that corresponds to class_label=3, which does not exist) will never be activated (the model will learn to never predict it).

In summary, this should work fine.

Ideally, you would change the ground truth, switching the value "4" with "3". And then you would set in the modelConfig, number-of-classes = 4. But hopefully, even in your current setup, the model should learn simply to never predict class=3.

I hope this answers it.

**Comment by asmaharry** [comment]
Thank you. I will try it as it is (with 5 classes and label 3 is missing)
then will let you know about the result.

**Comment by asmaharry** [comment]
hi,
I have 285*4 nii mri images and size of each image is 240*240*155. I am trying to train deepmedic on this data on gpu but it takes alot of time. untill now 48 hours has been passed and  only 2 epochs is completed. how to make training fast?
Thank you.

**Comment by Kamnitsask** [comment]
Hi,

It seems that the original issue has been resolved so I will close this one, to keep record clean.

If you encounter a new issue, please open a new issue ticket with a corresponding title etc.

(note: no, this slow down is not normal. Please double check that you are running your process correctly on the GPU and not the CPU (eg via nvidia-smi). After that, you can consult the logfile to check where the delays are happening and let us know (try 'grep "TIMING" ./path-to-logfile.txt'. )
